File: javascript/packages/flyql/src/tokenize.js
import { Parser as QueryParser, CharType as QueryCharType } from './core/index.js'
import { Parser as ColumnsParser, CharType as ColumnsCharType } from './columns/index.js'

const NUMERIC_RE = /^-?\d+(\.\d+)?([eE][+-]?\d+)?$/

function isNumericToken(s) {
    return NUMERIC_RE.test(s)
}

function upgradeQueryValue(text) {
    if (text === 'true' || text === 'false') return QueryCharType.BOOLEAN
    if (text === 'null') return QueryCharType.NULL
    if (isNumericToken(text)) return QueryCharType.NUMBER
    if (text.length > 0 && (text[0] === "'" || text[0] === '"')) return QueryCharType.STRING
    return QueryCharType.COLUMN
}

/**
 * Pure tokenizer: groups consecutive parser typed chars of the same CharType
 * into tokens with the shape { text, type, start, end }. Applies the canonical
 * VALUE upgrade in query mode. Appends a trailing ERROR token if the parser
 * halted before consuming the full input.
 *
 * @param {string} text - FlyQL expression
 * @param {object} [options]
 * @param {'query'|'columns'} [options.mode='query']
 * @returns {Array<{text: string, type: string, start: number, end: number}>}
 */
export function tokenize(text, options = {}) {
    const mode = options.mode || 'query'
    if (!text) return []

    let typedChars
    if (mode === 'columns') {
        const parser = new ColumnsParser({ transformers: true })
        parser.parse(text, false, false)
        typedChars = parser.typedChars || []
    } else {
        const parser = new QueryParser()
        parser.parse(text, false, false)
        typedChars = parser.typedChars || []
    }

    const tokens = []
    let curText = ''
    let curType = null
    let curStart = 0

    for (const [char, charType] of typedChars) {
        if (curType === null) {
            curText = char.value
            curType = charType
            curStart = char.pos
        } else if (charType === curType) {
            curText += char.value
        } else {
            tokens.push({
                text: curText,
                type: curType,
                start: curStart,
                end: curStart + curText.length,
            })
            curText = char.value
            curType = charType
            curStart = char.pos
        }
    }
    if (curType !== null) {
        tokens.push({
            text: curText,
            type: curType,
            start: curStart,
            end: curStart + curText.length,
        })
    }

    if (mode === 'query') {
        for (const token of tokens) {
            if (token.type === QueryCharType.VALUE) {
                token.type = upgradeQueryValue(token.text)
            }
        }
    }

    const consumed = tokens.length > 0 ? tokens[tokens.length - 1].end : 0
    if (consumed < text.length) {
        tokens.push({
            text: text.slice(consumed),
            type: mode === 'columns' ? ColumnsCharType.ERROR : QueryCharType.ERROR,
            start: consumed,
            end: text.length,
        })
    }

    return tokens
}

File: golang/tokenize.go
package flyql

import (
	"errors"
	"regexp"
	"unicode/utf8"
)

// Token is the cross-language token shape emitted by Tokenize.
// Fields mirror the JavaScript `{ text, type, start, end }` shape.
type Token struct {
	Text  string
	Type  string
	Start int
	End   int
}

var numericRe = regexp.MustCompile(`^-?\d+(\.\d+)?([eE][+-]?\d+)?$`)

func isNumeric(s string) bool {
	return numericRe.MatchString(s)
}

func upgradeQueryValue(text string) string {
	switch {
	case text == "true" || text == "false":
		return CharTypeBoolean
	case text == "null":
		return CharTypeNull
	case isNumeric(text):
		return CharTypeNumber
	case len(text) > 0 && (text[0] == '\'' || text[0] == '"'):
		return CharTypeString
	default:
		return CharTypeColumn
	}
}

// Tokenize groups the parser's per-character TypedChars output into token
// runs with shape {Text, Type, Start, End}, applies the canonical VALUE
// upgrade in query mode, and appends a trailing ERROR token when the
// parser halts before consuming the full input. Columns mode is only
// available in the JavaScript package and returns an error here.
func Tokenize(text, mode string) ([]Token, error) {
	if mode == "" {
		mode = "query"
	}
	if mode != "query" {
		return nil, errors.New("columns mode is only available in the JavaScript package")
	}
	if text == "" {
		return []Token{}, nil
	}

	p := NewParser()
	_ = p.Parse(text)
	typed := p.TypedChars

	tokens := make([]Token, 0)
	if len(typed) > 0 {
		curText := string(typed[0].Value)
		curType := typed[0].Type
		curStart := typed[0].Pos
		for i := 1; i < len(typed); i++ {
			c := typed[i]
			if c.Type == curType {
				curText += string(c.Value)
			} else {
				tokens = append(tokens, Token{
					Text:  curText,
					Type:  curType,
					Start: curStart,
					End:   curStart + utf8.RuneCountInString(curText),
				})
				curText = string(c.Value)
				curType = c.Type
				curStart = c.Pos
			}
		}
		tokens = append(tokens, Token{
			Text:  curText,
			Type:  curType,
			Start: curStart,
			End:   curStart + utf8.RuneCountInString(curText),
		})
	}

	for i := range tokens {
		if tokens[i].Type == CharTypeValue {
			tokens[i].Type = upgradeQueryValue(tokens[i].Text)
		}
	}

	consumed := 0
	if len(tokens) > 0 {
		consumed = tokens[len(tokens)-1].End
	}
	// consumed is a code-point offset, so the trailing slice and total length
	// must be measured in code points too (text indexes bytes natively).
	runes := []rune(text)
	if consumed < len(runes) {
		tokens = append(tokens, Token{
			Text:  string(runes[consumed:]),
			Type:  CharTypeError,
			Start: consumed,
			End:   len(runes),
		})
	}

	return tokens, nil
}
